File: 2.0/WearOS/WatchSprings-2.0/app/src/main/java/copyright/markmorris/allrightsreserved/watchsprings/screens/models/StatusViewModel.kt
// WatchSprings 2.0 © 2025 Mark S. Morris. All rights reserved.
//
// https://www.watchsprings.com
//
// WatchSprings is Fan Content pursuant to the Epic Games Fan Content Policy. Portions of the materials used are trademarks and/or copyrighted works of Epic Games, Inc.

package copyright.markmorris.allrightsreserved.watchsprings

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class StatusViewModel : ViewModel() {

    private val _fortniteComponents = MutableStateFlow<List<StatusComponent>>(emptyList())
    val fortniteComponents = _fortniteComponents.asStateFlow()

    private val _isFortniteOnline = MutableStateFlow(false)
    val isFortniteOnline = _isFortniteOnline.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading = _isLoading.asStateFlow()

    private val _error = MutableStateFlow<String?>(null)
    val error = _error.asStateFlow()

    fun refreshStatus() {
        if (_fortniteComponents.value.isNotEmpty()) return

        viewModelScope.launch {
            _isLoading.value = true
            try {
                val result = MainActivity.fortniteAPI.getServerStatus()
                val components = result.components

                val fortniteGroup = components.find { it.name == "Fortnite" && it.group }

                val fortniteChildren = if (fortniteGroup != null) {
                    _isFortniteOnline.value = fortniteGroup.status == "operational"
                    components.filter { it.groupId == fortniteGroup.id }
                } else {
                    _isFortniteOnline.value = false
                    emptyList()
                }

                _fortniteComponents.value = fortniteChildren

            } catch (e: Exception) {
                _error.value = "Failed to load status: ${e.message}"
                println("Error: ${e.message}")
            } finally {
                _isLoading.value = false
            }
        }
    }
}

File: 2.0/watchOS/WatchSprings-2.0/WatchSprings Watch App/screens/models/StatusViewModel.swift
// WatchSprings 2.0 © 2025 Mark S. Morris. All rights reserved.
//
// https://www.watchsprings.com
//
// WatchSprings is Fan Content pursuant to the Epic Games Fan Content Policy. Portions of the materials used are trademarks and/or copyrighted works of Epic Games, Inc.

import Foundation
import Kingfisher

func print(_ items: Any...)
{
    #if DEBUG
    Swift.print("DEBUG " + Date().description,items[0])
    #endif
}

@Observable
class StatusViewModel
{
    var fortniteComponents: [StatusComponent] = []
    var isFortniteOnline: Bool = false
    var isLoading: Bool = false
    var error: String?
    
    init()
    {
        ImageCache.default.clearCache()
    }

    func refreshStatus() async
    {
        guard fortniteComponents.isEmpty else { return }

        isLoading = true
        defer { isLoading = false }

        do {
            let result = try await FortniteApiClient.instance.getServerStatus()
            let components = result.components

            // Locate the Fortnite group component
            if let fortniteGroup = components.first(where: { $0.name == "Fortnite" && $0.group }) {
                isFortniteOnline = fortniteGroup.status == "operational"
                fortniteComponents = components.filter { $0.groupId == fortniteGroup.id }
            } else {
                isFortniteOnline = false
                fortniteComponents = []
            }
        } catch {
            self.error = "Failed to load status: \(error.localizedDescription)"
            print("Error fetching status: \(error)")
        }
    }
}

